dao-voting/src/main.sw

```diff
@@ -0,0 +1,272 @@
+contract;
+
+dep abi;
+dep data_structures;
+dep errors;
+
+use std::{
+    address::Address,
+    assert::require,
+    block::height,
+    chain::auth::{AuthError, msg_sender},
+    context::{call_frames::msg_asset_id, msg_amount, this_balance},
+    contract_id::ContractId,
+    identity::Identity,
+    result::*,
+    revert::revert,
+    storage::StorageMap,
+    token::{transfer_to_output, force_transfer_to_contract},
+};
+
+use abi::DaoVoting;
+use data_structures::{CallData, Proposal};
+use errors::Error;
+
+storage {
+    gov_token: ContractId,
+    voting_period: u64,
+    approval_percentage: u64,
+    proposals: StorageMap<u64,
+    Proposal>, proposal_count: u64,
+    // The amount of governance tokens a user has deposited
+    balances: StorageMap<Identity,
+    u64>,
+    // The amount of votes a user has used on a proposal
+    votes: StorageMap<(Identity, u64),
+    u64>, state: u64, 
+}
+
+impl DaoVoting for Contract {
+    /// Initialize the dao with the governance token, voting parameters, and the proposal.
+    ///
+    /// # Panics
+    ///
+    /// The function will panic when:
+    /// - The constructor is called more than once
+    /// - The voting period is not greater than 0
+    /// - The approval percentage is not greater than 0
+    #[storage(read, write)]
+    fn constructor(gov_token: ContractId) -> bool {
+        require(storage.state == 0, Error::CannotReinitialize);
+
+        storage.gov_token = gov_token;
+        storage.proposal_count = 0;
+        storage.state = 1;
+
+        true
+    }
+
+    /// Add proposal to be voted on
+    ///
+    /// # Panics
+    ///
+    /// The function will panic when:
+    /// - The constructor has not been called to initialize
+    /// - The voting period is 0
+    /// - The approval percentage is 0
+    #[storage(read, write)]
+    fn add_proposal(voting_period: u64, approval_percentage: u64, proposal_data: CallData) -> bool {
+        require(storage.state == 1, Error::NotInitialized);
+        require(voting_period > 0, Error::PeriodCannotBeZero);
+        require(approval_percentage > 0, Error::ApprovalPercentageCannotBeZero);
```

Comment:
Does the contract handle cases where someone enters an `approval_percentage` greater than 100? Is this dangerous?

Reply:
good catch!  It doesn't so if the approval percentage was greater than 100 it would not be able to get approved.